src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -1640,7 +1663,7 @@ private function hasReference($id, array $arguments, $deep = false, array &$visi
                     return true;
                 }
 
-                if (!$deep || isset($visited[$argumentId]) || !isset($this->circularReferences[$id][$argumentId])) {
+                if (!$deep || isset($visited[$argumentId]) || !isset($this->circularReferences[$argumentId])) {
```

Comment:
`$this->circularReferences` tells if an id is involved in a circular loop, but checking for both $id & $argumentId is broken with loops that involve more than two services.